Updated handling of config values reading bad configuration

rolfyone opened the pull request:
The constants that got moved have caused issues in remote config, and now in local config.

RemoteConfig was loading the local config to work around the issue but it was in the wrong order, presets load first, then config.

LocalConfig could potentially do this with a CLI flag but shouldn't be default behaviour IMO, so i've left local config failing.

The error handling has been updated to report all fields not just the first field that fails validation. The error will look something like:
```
java.lang.IllegalArgumentException: The specified network configuration had missing or invalid values for constants GOSSIP_MAX_SIZE, ATTESTATION_SUBNET_PREFIX_BITS, RESP_TIMEOUT, ATTESTATION_SUBNET_EXTRA_BITS, MAXIMUM_GOSSIP_CLOCK_DISPARITY, MAX_CHUNK_SIZE, ATTESTATION_PROPAGATION_SLOT_RANGE, MESSAGE_DOMAIN_INVALID_SNAPPY, TTFB_TIMEOUT, EPOCHS_PER_SUBNET_SUBSCRIPTION, MESSAGE_DOMAIN_VALID_SNAPPY, ATTESTATION_SUBNET_COUNT, MIN_EPOCHS_FOR_BLOCK_REQUESTS, MAX_REQUEST_BLOCKS
```

fixes #7555

## Documentation

- [X] I thought about documentation and added the `doc-change-required` label to this PR if updates are required.

## Changelog

- [X] I thought about adding a changelog entry, and added one if I deemed necessary.

rolfyone committed:
Defaulted config attributes that were moved from constants

The constants that got moved have caused issues in remote config, and now in local config.

This change defaults the values, which means if they're not supplied it'll silently succeed assuming the values from old constant values.

I've reverted the initial remote config change, as it didn't really address the local loading issue mentioned in #7555, and it was potentially also applying in the incorrect order.

`MIN_EPOCHS_FOR_BLOCK_REQUESTS` is different on mainnet and minimal, so that will be using the mainnet value by default.

fixes #7555

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
updated changelog.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
change update messages down to trace

Signed-off-by: Paul Harris <paul.harris@consensys.net>

jtraglia commented on ethereum/spec/src/main/java/tech/pegasys/teku/spec/config/builder/SpecConfigBuilder.java:
```diff
@@ -102,21 +104,21 @@ public class SpecConfigBuilder {
+  private Integer maximumGossipClockDisparity = 500;
+  private Bytes4 messageDomainInvalidSnappy = Bytes4.fromHexString("0x00000000");
+  private Bytes4 messageDomainValidSnappy = Bytes4.fromHexString("0x01000000");
   private Integer subnetsPerNode;
```
Should `subnetsPerNode` have a default?

rolfyone replied:
this should probably follow the same pattern (whatever is decided below) i was mapping against another PR and missed this one...

rolfyone committed:
Merge branch 'master' into fix-remote-loader

rolfyone committed:
refactored to put defaults in the reader rather than the builder.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into fix-remote-loader

rolfyone committed:
Merge remote-tracking branch 'upstream/master' into fix-remote-loader

rolfyone committed:
Now back to automatically loading the defaults for remote configuration

For local configuration, will still fail to load, but will provide all of the fields failing validation, not just the first.

Arguably we could have a CLI arg that loads default config for local resources also, but Im not sure that it should be the default, as we'd be hiding actual configuration issues and never be made aware.

Error will be
```
java.lang.IllegalArgumentException: The specified network configuration had missing or invalid values for constants GOSSIP_MAX_SIZE, ATTESTATION_SUBNET_PREFIX_BITS, RESP_TIMEOUT, ATTESTATION_SUBNET_EXTRA_BITS, MAXIMUM_GOSSIP_CLOCK_DISPARITY, MAX_CHUNK_SIZE, ATTESTATION_PROPAGATION_SLOT_RANGE, MESSAGE_DOMAIN_INVALID_SNAPPY, TTFB_TIMEOUT, EPOCHS_PER_SUBNET_SUBSCRIPTION, MESSAGE_DOMAIN_VALID_SNAPPY, ATTESTATION_SUBNET_COUNT, MIN_EPOCHS_FOR_BLOCK_REQUESTS, MAX_REQUEST_BLOCKS
```
which is arguably harder to read, but doesnt force you to fix one, rinse repeat until you've got a valid configuration.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
changelog.

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge branch 'master' into fix-remote-loader

rolfyone committed:
Merge branch 'master' into fix-remote-loader

rolfyone committed:
Merge branch 'master' into fix-remote-loader

rolfyone committed:
review feedback.

Also improved a test case to allow us to see multiple fields returned (without listing the actual fields)

Signed-off-by: Paul Harris <paul.harris@consensys.net>

rolfyone committed:
Merge branch 'master' into fix-remote-loader
